docs/proposals/20220210-md-node-label-sync.md

```diff
@@ -0,0 +1,165 @@
+
+#### Story 1
+
+As a cluster admin/user, I would like a declarative and secure means by which to assign roles to my nodes. For example, when I do `kubectl get nodes`, I want the ROLES column to list my assigned roles for the nodes.
+
+#### Story 2
+
+As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.
+
+### Implementation Details/Notes/Constraints
+
+#### Label propogation from MachineDeployment to Machine
+
+Utilize CAPI metadata Propagation. CAPI supports [metadata propagation](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html?highlight=propagation#metadata-propagation) across core API resources. For the MachineDeployment types, template labels propagate to MachineSets top-level and MachineSets template metadata:
+```
+MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
+```
+Similarly, on MachineSet, template labels propagate to Machines:
+```
+MachineSet.spec.template.metadata.labels => Machine.labels
+```
+
+As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. CAPI will be updated to instead ignore updates of labels that fall within the CAPI prefix. There is precedence for this with the handling of `MachineDeploymentUniqueLabel`. 
+
+#### Synchronization of CAPI Labels
+
+Labels that fall under the CAPI specific prefix `node.cluster.k8s.io/*` will need to be kept in sync between the Machine and Node objects. Synchronization must handle these two scenarios: 
+(A) Label is added/removed on a Machine and must be added/removed on the corresponding Node.
+(B) Label is added/removed on the Node and must be removed/re-added on the Node to bring it in sync with the labels on Machine.
+
+
+#### Machine Controller Changes
+
+On each Machine reconcile event, the machine controller will fetch the corresponding Node (i.e. Machine.Status.NodeRef). If the Machine has labels specified, then the controller will ensure those labels are applied to the Node. 
+
+Similarly, today the machine controller watches Nodes within the workload cluster, where changes to the Node trigger a reconcile of its corresponding Machine object. This should allow us to capture changes made directly to the labels on the Node (e.g. user accidently deletes a label that falls under CAPI's managed prefix).
+
+#### One-time Apply of Standard Kubernetes Labels
+
+The machine controller will apply the standard kubernetes labels, if specified, to the Node immediately after the Node enters the Ready state (ProviderID is set). We will enforce this one-time application by making the labels immutable via a validating webhook. 
```

Comment:
@fabriziopandini This could be an interesting edge case if we choose to in-place update all labels from MD=>MS=>M

The plan was to only apply those labels one time on a node and then don't change them later. But if a label change doesn't trigger an entire rollout I'm not sure how we can rollout those labels from a MD to nodes

Reply:
Discussed it. We can solve this either with rolloutAfter or document the limitation